### PR TITLE
Make Go runfiles library repo mapping aware

### PR DESCRIPTION
**Pull request by fmeum**
Implements https://github.com/bazelbuild/proposals/blob/main/designs/2022-07-21-locating-runfiles-with-bzlmod.md for rules_go.

Work towards bazelbuild/bazel#16124

### CODE REVIEW

**Comment by fmeum** [comment]
@phst I implemented my runfiles location proposal for Go, based on our import of your runfiles library. If you have the time, it would be great if you could take a look.

**Comment by fmeum** [comment]
@Wyverald Would you be comfortable reviewing this Go PR, at least from the Bzlmod angle?

**Review by Wyverald — APPROVED** [review]
LGTM!

**Comment by fmeum** [comment]
@linzhp @achew22 Could one of you take a brief look at the new API this introduces (`RunfilesFrom`, `CurrentRepository`) and whether it is sufficiently idiomatic Go? I pushed a small additional commit to bring `CurrentRepository`'s arg handling in line with `runtime.Caller`.

File: go/runfiles/global.go [hunk 1]
```diff
@@ -42,6 +50,26 @@ func Env() ([]string, error) {
...
 }
 
+var legacyExternalGeneratedFile = regexp.MustCompile(`^bazel-out[/][^/]+/bin/external/([^/]+)/`)
+var legacyExternalFile = regexp.MustCompile(`^external/([^/]+)/`)
+
+// CurrentRepository returns the canonical name of the Bazel repository that
+// contains the source file of the caller of this function with skip frames
+// skipped. For example, passing 0 will return the name of the repository that
+// contains the caller of CurrentRepository.
+func CurrentRepository(skip int) string {
```
Comment (linzhp) [inline go/runfiles/global.go:60]: Can you create functions `CallerRepository` and `CurrentRepository`?

File: go/runfiles/manifest.go [hunk 2]
```diff
@@ -28,13 +28,23 @@ import (
...
+	repoMappingPath, err := m.path(repoMappingRlocation)
+	// If Bzlmod is disabled, the repository mapping manifest isn't created, so
+	// it is not an error if it is missing.
+	if err == nil {
+		repoMapping, err = parseRepoMapping(repoMappingPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &Runfiles{m, manifestFileVar + "=" + string(f), repoMapping, string(sourceRepo)}, nil
```
Comment (linzhp) [inline go/runfiles/manifest.go:47]: Duplicate code with Directory.new?

**Review by linzhp — COMMENTED** [review]
Some Go stylish comments without understanding the logic :)

File: go/runfiles/manifest.go (hunk 2 above)
Comment (fmeum) [reply go/runfiles/manifest.go]: I extracted and cleaned up the common code.

File: go/runfiles/directory.go [hunk 3]
```diff
@@ -21,8 +21,18 @@ import "path/filepath"
...
+	var repoMapping map[repoMappingKey]string
+	repoMappingPath, err := d.path(repoMappingRlocation)
+	// If Bzlmod is disabled, the repository mapping manifest isn't created, so
+	// it is not an error if it is missing.
+	if err == nil {
+		repoMapping, err = parseRepoMapping(repoMappingPath)
+		if err != nil {
+			return nil, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (fmeum) [reply go/runfiles/directory.go]: The refactoring also reduced the nesting level.

File: go/runfiles/global.go (hunk 1 above)
Comment (fmeum) [reply go/runfiles/global.go]: I did, is this what you expected?

File: go/runfiles/global.go [hunk 4]
```diff
@@ -42,6 +50,33 @@ func Env() ([]string, error) {
 	return r.Env(), nil
 }
 
+var legacyExternalGeneratedFile = regexp.MustCompile(`^bazel-out[/][^/]+/bin/external/([^/]+)/`)
+var legacyExternalFile = regexp.MustCompile(`^external/([^/]+)/`)
+
+// CurrentRepository returns the canonical name of the Bazel repository that
+// contains the source file of the caller of this function.
+func CurrentRepository() string {
+	return CallerRepository(1)
```
Comment (linzhp) [inline go/runfiles/global.go:59]:
```suggestion
	return CallerRepository(0)
```
Should this be 0, according to the doc of `CallerRepository`?

**Review by linzhp — COMMENTED** [review]
All current calls to `CallerRepository` pass either 0 or 1. I am thinking about making `CallerRepository(0)` become `CurrentRepository()` and `CallerRepository(1)` become `CallerRepository()` or `ParentRepository()` without any parameters. This easier to read this way. It's not as flexible but we can always add a more flexible function like `CallerRepositoryWithSkips()` later on when there is a need.

**Comment by fmeum** [comment]
@linzhp The numerical argument is confusing enough that I like the idea to get rid of it. Done.

**Review by linzhp — APPROVED** [review]

**Pull request merged**